src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageService.cs

```diff
@@ -0,0 +1,186 @@
+            IConnectionFactory factory = new ConnectionFactory();
+            factory.UserName = _userName;
+            factory.Password = _password;
+            return factory.CreateConnection();
+        }
+
+        /// <summary>
+        /// Publish the message
+        /// </summary>
+        /// <param name="messageType"></param>
+        /// <param name="message"></param>
+        /// <param name="delay"></param>
+        /// <param name="cancellationToken"></param>
+        /// <remarks>RabbitMQ has an upper limit of 2GB for messages.BasicPublish blocking AMQP operations.
+        /// The rule of thumb is: avoid sharing channels across threads.
+        /// Publishers in your application that publish from separate threads should use their own channels.
+        /// The same is a good idea for consumers.</remarks>
+        /// <returns></returns>
+        public override async Task PublishAsync(Type messageType, object message, TimeSpan? delay = null,
+            CancellationToken cancellationToken = default(CancellationToken)) {
+            if (message == null)
+                return;
+            _logger.Trace("Message Publish: {messageType}", messageType.FullName);
+
+            if (delay.HasValue && delay.Value > TimeSpan.Zero) {
+                await AddDelayedMessageAsync(messageType, message, delay.Value).AnyContext();
+                return;
+            }
+            await Run.WithRetriesAsync(() => Publish(messageType, message, cancellationToken), logger: _logger, cancellationToken: cancellationToken).AnyContext();
+        }
+
+        private async Task Publish(Type messageType, object message, CancellationToken cancellationToken = default(CancellationToken)) {
+            IBasicProperties basicProperties = _publisherChannel.CreateBasicProperties();
+            basicProperties.Persistent = _persistent;
+            basicProperties.Expiration = _defaultMessageTimeToLive.Milliseconds.ToString();
+
+            var data = await _serializer.SerializeAsync(new MessageBusData {
+                Type = messageType.AssemblyQualifiedName,
+                Data = await _serializer.SerializeToStringAsync(message).AnyContext()
+            }).AnyContext();
+            // The publication occurs with mandatory=false
+            _publisherChannel.BasicPublish(_exchangeName, _routingKey, basicProperties, data);
+        }
+
+        /// <summary>
+        /// Subscribe for the message
+        /// </summary>
+        /// <typeparam name="T">Type of the subscriber who wants to be notified of the callback</typeparam>
+        /// <param name="handler">callback handler</param>
+        /// <param name="cancellationToken"></param>
+        public override void Subscribe<T>(Func<T, CancellationToken, Task> handler,
+            CancellationToken cancellationToken = default(CancellationToken)) {
+            var consumer = new EventingBasicConsumer(_subscriberChannel);
+            consumer.Received += async (o, e) => {
```

Comment:
@niemyjski If the object gets disposed then there is no need to worry about unsubscription. I have tested it myself where once the Dispose is called where we are closing the connection and channels, the event handler never gets called.